src/s3/s3_object.rs

```diff
@@ -0,0 +1,131 @@
+use std::str::FromStr;
+
+use anyhow::{bail, Context, Error, Result};
+use serde::{Deserialize, Serialize};
+use url::Url;
+
+/// A bucket key pair for a S3Object, with conversion from S3 urls.
+#[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
+pub struct S3Object {
+    /// The bucket the object is in.
+    pub bucket: String,
+    /// The key the in the bucket for the object.
+    pub key: String,
+}
+
+impl S3Object {
+    /// Create a new [S3Object] using anything which can be
+    /// treated as [&str].  Any leading `/` will be trimmed from
+    /// the key.  No validation is done against the bucket or key
+    /// to ensure they meet the AWS requirements.
+    pub fn new(bucket: impl AsRef<str>, key: impl AsRef<str>) -> Self {
+        S3Object {
+            bucket: bucket.as_ref().to_owned(),
+            key: key.as_ref().trim_start_matches('/').to_owned(),
+        }
+    }
+}
+
+/// Convert from an [Url] into a [S3Object]. The scheme
+/// must be `s3` and the `path` must not be empty.
+impl TryFrom<Url> for S3Object {
+    type Error = Error;
+
+    fn try_from(value: Url) -> Result<Self, Self::Error> {
+        if value.scheme() != "s3" {
+            bail!("S3 URL must have a scheme of s3")
+        }
+        let bucket = value.host_str().context("S3 URL must have host")?;
+        let key = value.path();
+
+        if key.is_empty() {
+            bail!("S3 URL must have a path")
+        }
+        Ok(S3Object::new(bucket, key))
+    }
+}
+
+/// Convert from a [&str] into a [S3Object].
+/// The [&str] must be a valid `S3` [Url].
+impl TryFrom<&str> for S3Object {
+    type Error = Error;
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        value.parse::<Url>()?.try_into()
+    }
+}
+
+/// Covert from [String] into a [S3Object].
+/// The [String] must be a valid `S3` [Url].
+impl TryFrom<String> for S3Object {
+    type Error = Error;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        value.parse::<Url>()?.try_into()
+    }
+}
+
+/// Covert from [&str] into a [S3Object].
+/// The [&str] must be a valid `S3` [Url].
+impl FromStr for S3Object {
+    type Err = Error;
+
+    fn from_str(value: &str) -> Result<Self, Self::Err> {
+        value.parse::<Url>()?.try_into()
+    }
+}
+
+/// Converts from a [S3Object] into a [Url].
+/// If the [S3Object] holds an invalid path or
+/// domain the conversion will fail.
+impl TryFrom<S3Object> for Url {
+    type Error = url::ParseError;
+    fn try_from(obj: S3Object) -> std::result::Result<Self, Self::Error> {
+        Url::parse(&format!("s3://{}/{}", obj.bucket, obj.key))
+    }
+}
+
+/// Converts from a [S3Object] into a [Url].
+/// If the [S3Object] holds an invalid path or
+/// domain the conversion will fail.
+impl TryFrom<&S3Object> for Url {
+    type Error = url::ParseError;
+    fn try_from(obj: &S3Object) -> std::result::Result<Self, Self::Error> {
+        Url::parse(&format!("s3://{}/{}", obj.bucket, obj.key))
+    }
+}
```

Comment:
If we have an implementation for `TryFrom<S3Object>`, can rely on the compiler to automatically handle the `TryFrom<&S3Object>` case? My intuition is that it should work, but I'm not certain 🤔

Reply:
My understanding is that the compiler does no implement this and neither is is defined by a generic type. Simple example shows this being the case for `From`.  I assume there must be cases where having this implementation would be disadvantageous. 

```rust

struct Foo();
struct Bar();

impl From<Bar> for Foo {

    fn from(bar: Bar) -> Self {
        Foo()
    }
}


fn main() {

    let b = Bar();
    let _: Foo = b.into();
    
    let b_ref = &Bar();
    let _: &Foo = b_ref.into();
   
}
```

This fails with 
```rust
error[[E0277]](https://doc.rust-lang.org/stable/error-index.html#E0277): the trait bound `&Foo: From<&Bar>` is not satisfied
  --> src/main.rs:19:19
   |
19 |     let _: &Foo = b_ref.into();
   |                   ^^^^^ ---- required by a bound introduced by this call
   |                   |
   |                   the trait `From<&Bar>` is not implemented for `&Foo`
   |
   = help: the trait `From<Bar>` is implemented for `Foo`
```

[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=10d0d6e5ea8684cfa56848715a6d6347)